### ISSUE TITLE
[BUG] `print(None)` doesn't work anymore

### ISSUE THREAD

**Issue by soraros** [modular/max#3319]
### Bug description

As title.

### Steps to reproduce

```mojo
fn main():
  print(None)  # error: invalid call to 'print': callee with non-empty variadic pack argument
               # expects 0 positional operands, but 1 was specified
```

### System information

```shell
Mojo 2024.7.2705 on Docker, Intel Mac
```

**Comment by soraros** [comment]
@JoeLoser I don't think it's fixed (`mojo 2024.9.2205 (3407615a)`).

```mojo
fn main():
  print(None)        # same error
  print(NoneType())  # works
```

**Comment by JoeLoser** [comment]
Agreed, this looks like it was improperly closed from the work in [https://github.com/modularml/mojo/pull/3387](https://github.com/modularml/mojo/pull/3387), but that only fixes `NoneType` and not `None`.  @connorg `None` is still special in the compiler in this regard, right?  I.e. it's an MLIR-first type and so the compiler isn't decorating it with the `Formattable` trait or anything of the like.

**Comment by soraros** [comment]
@JoeLoser I don't think this is actually fixed. `print(None)` still fails.

**Comment by rd4com** [comment]
Could a possible solution be to make `None` go to a `non-materializable` type that materializes into `NoneType` ?

**Comment by soraros** [comment]
@rd4com I was the thinking the same thing. `None` is essentially `NoneLiteral`.

**Comment by JoeLoser** [comment]
Re-opening as the issue still exists.  We have `NoneType` in the library, but there's more work to do in the compiler here to make this work.  `None` in the compiler is essentially just the `kgen.none` MLIR type.  We can't arbitrarily add the `Writable` trait (which `print` requires in the library) to MLIR types.  We'd need to have `None`  be more of a first-class library type that way we could add said trait to make this work.  @ConnorGray may remember some of the history here on this.

**Comment by ConnorGray** [comment]
(Posting this comment again as I got an error the first time, apologies if its ends up being a duplicate.)

Yes, that's my understanding as well. If I recall this design space correctly, we can't easily fix this without compiler work that would change how the None identifier works.

One "obvious" solution would be to promote None to a Mojo-backed type, i.e. rename struct NoneType to None. That would allow us to implement methods on None.

But it would leave us with a different problem, because the None identifier is both a type and a value, and if I understand correctly, has special handling by the compiler because of that.

E.g. as one example, field-less Mojo struct can't be initialized just by doing Foo. At the very least you have to write Foo().

**Comment by soraros** [comment]
@ConnorGray Can we treat `None` as a constant literal of some non-materialisable `LiteralNone` type, similar to how we do with `IntLiteral`, etc.? Though the fact that `None` doubles as a type literal doesn't help.
